### PR TITLE
Webpack build system (POC)

### PR DESCRIPTION
**Pull request by cannoneyed**
This PR is a POC for building magenta.js using webpack. 

Basically, all attempts to get rollup working were foiled by some third-party dependencies (mainly `midiconver` and `protobufjs`) not playing very nicely with rollup, and there doesn't seem to be a way around it for the time being.

Webpack offers a lot of similar benefits to rollup (more modern, extensible), plus it allows for a much nicer development workflow around the demos (watch mode and live reload). There's really a lot going on in this PR, and I'd like to split it up into subsequent smaller PRs that are each concerned with one step of the process, but here's the basic outline of what's going on here:

- the demos and lib are now built using webpack rather than a browserify-centered shell script.
- demos are now organized via a demo-per-directory structure
- shell scripts moved to their own directory
- npm commands updated

Hopefully everything works as-is, it would be nice to do a quick sanity check on all of the machinery around building / publishing. There's still a lot of potential work to be done around splitting out libs / modularizing the build setup a bit, but I think this at least gives us some modern build tooling to start addressing those ideas.

### CODE REVIEW

**Comment by notwaldorf** [comment]
Not super sure the build is correct. I added this super basic `index.html` in `music/dist` (after running a build and producing a `magentamusic.js`:

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <script src="./magentamusic.js"></script>
</head>
<body>
</body>
<script>
  const player = new mm.Player();
</script>
</html>
```

and when running it I get" (index):11 Uncaught ReferenceError: mm is not defined"

**Comment by notwaldorf** [comment]
Also, looking at the `magentamusic.js` created, it seems in the same order of magnitude as what we have now, which is an unminified. The webpack branch creates a file that's 3.2Mb, but the one on master is 3.4Mb (so only a bit bigger). I think we can get a bunch of gains if we actually minify the webpack produced file -- can you look into what that would look like? (like, actually producing a `magentamusic.min.js`, and not just relying on the CDN to minify it for us)

File: music/demos/index.html [hunk 1]
```diff
@@ -37,10 +37,10 @@ <h2>Models</h2>
   <section>
     <h3>
       <ul>
-        <li><a href="music_rnn.html">Continue a NoteSequence using MusicRNN</a></li>
-        <li><a href="music_vae.html">Generate new NoteSequences using MusicVAE</a></li>
-        <li><a href="multitrack.html">Multitrack MusicVAE</a></li>
-        <li><a href="transcription.html">Piano transcription with Onsets and Frames</a></li>
+        <li><a href="music_rnn/music_rnn.html">Continue a NoteSequence using MusicRNN</a></li>
```
[the remark this answers is not in the document]
Comment (cannoneyed) [reply music/demos/index.html]: Hmm, yeah i'm not 100% convinced either, it just seemed like in a world where there may be more functionality per demo or more demos, then having a mega-folder gets a bit unwieldy. But I don't think it's too important.

File: music/demos/visualizer/visualizer.ts [hunk 2]
```diff
@@ -15,11 +15,11 @@
  * limitations under the License.
  */
 
-import * as mm from '../src/index';
+import * as mm from "../../src/index";
```
[the remark this answers is not in the document]
Comment (cannoneyed) [reply music/demos/visualizer/visualizer.ts]: Bah, yeah I thought I caught all of this - my bad (VSCode had autoformat turned on with no formatting options...) - will fix

File: music/package.json [hunk 3]
```diff
@@ -29,25 +29,30 @@
...
     "typedoc": "^0.11.1",
     "typedoc-plugin-sourcefile-url": "^1.0.3",
-    "typescript": "^2.7.2"
+    "typescript": "^2.7.2",
+    "webpack": "^4.24.0",
+    "webpack-cli": "^3.1.2"
   },
   "scripts": {
-    "prepublish": "yarn lint && yarn test && yarn build && yarn doc && yarn bundle",
+    "prepublish": "in-publish && yarn lint && yarn test && yarn build && yarn doc && yarn bundle || not-in-publish",
```
[the remark this answers is not in the document]
Comment (cannoneyed) [reply music/package.json]: So there's an annoying "feature/bug" (depending on who you ask) where any npm/yarn install triggers `prepublish`, which means that we build/run tests/all the things whenever we add a new dependency, even if it's as simple as just trying it out in a feature branch. `in-publish` basically allows you to only run certain things when actually running the `publish` command

File: music/package.json [hunk 4]
```diff
@@ -29,25 +29,30 @@
...
-    "publish-demos": "yarn build-demos && mkdir -p ../docs/music/demos && cp demos/*.{js,html,mid,css} ../docs/music/demos",
-    "run-demos": "yarn build-demos && http-server demos/",
-    "proto": "sh compile-proto.sh",
-    "doc": "sh generate-docs.sh && yarn publish-demos"
+    "build-demos": "webpack --config ./webpack/demos.config.ts",
+    "publish-demos": "yarn build-demos && mkdir -p ../docs/music/demos && cp ./demos/**/*.{js,html} ../docs/music/demos && cp ./demos/*.{mid,css} ../docs/music/demos",
+    "run-demos": "yarn build-demos && yarn serve-demos",
+    "proto": "sh ./scripts/compile-proto.sh",
+    "docs": "sh ./scripts/generate-docs.sh && yarn publish-demos",
+    "serve-demos": "http-server demos/"
```
[the remark this answers is not in the document]
Comment (cannoneyed) [reply music/package.json]: 👍

**Comment by cannoneyed** [comment]
Re: minification - it's as simple as turning on a flag on the production webpack config file, but I actually wanted to check in with you first to see if this was something we wanted to do (noticed we were doing it through the CDN before)

And re: the build working, I'm gonna triple check everything and see what the deal is there... might be something isn't working on a fresh clone

**Comment by cannoneyed** [comment]
OK cool - just had a chance to look into the build issue and it makes a lot of sense. Are we expecting to add a global `mm` object by including the script? I guess for a CDN-delivered magenta.js this makes a lot of sense, but we're probably going to want to build it as a module for npm.

I'll update the build setup accordingly, but I'd love to chat about the distribution story so we can best decide what to build for where.

File: music/demos/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/demos/index.html]: i don't think the rate at which we're adding demos is high enough for this worry (we tend to add a new demo for each model, and models don't get dropped every day), so I think for now i'd prefer the all-demos-in-one folder approach. wdyt @adarob?

File: music/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/package.json]: ah, super cool, i hated the previous behaviour.

**Comment by notwaldorf** [comment]
> Are we expecting to add a global mm object by including the script? I guess for a CDN-delivered magenta.js this makes a lot of sense, but we're probably going to want to build it as a module for npm.

Yeah, i think we *definitely* want to export an `mm` global. This is what the current `magentamusic.js` file does, which means that if we don't continue exporting it, it's a huge breaking change and it might break all existing demos (if they use ^ and don't pin a specific version). When we do have module-enabled builds, i think they should have a different name (or maybe in a different folder) to keep the existing demos still ok

> Re: minification - it's as simple as turning on a flag on the production webpack config file, but I actually wanted to check in with you first to see if this was something we wanted to do (noticed we were doing it through the CDN before)
Don't know why we weren't doing it before (@adarob might know), but I think we should minify it ourselves (this is how we could've caught a few versions in advance that our build was broken and we weren't exactly shipping pure es5 anymore)

**Comment by adarob** [comment]
Yes, let's minify ourselves now. The reason we weren't doing it before was because I had/have no idea what I was/am doing.

**Comment by cannoneyed** [comment]
OK y'all, just had a free moment to get back into this work. Here's the latest status of the things we talked about:

- `yarn bundle` builds a minified library by default.
- `yarn bundle` now builds `src/lib.ts`, which exports the library as `mm` on the window object. 
- All of the demos are back to single files in the `demos` directory.

Sorry for the little hiatus - LMK how this is looking!

**Comment by notwaldorf** [comment]
As a side note, running `yarn prepublish` doesn't work for me (error: `/bin/sh: in-publish: command not found`)

File: music/package.json [hunk 5]
```diff
@@ -30,25 +30,31 @@
...
     "typedoc": "^0.11.1",
     "typedoc-plugin-sourcefile-url": "^1.0.3",
-    "typescript": "^2.7.2"
+    "typescript": "^2.7.2",
+    "webpack": "^4.24.0",
+    "webpack-cli": "^3.1.2"
   },
   "scripts": {
-    "prepublish": "yarn lint && yarn test && yarn build && yarn doc && yarn bundle",
+    "prepublish": "in-publish && yarn lint && yarn test && yarn build && yarn doc && yarn bundle || not-in-publish",
```
Comment (notwaldorf) [inline music/package.json]: typo, this should be `yarn docs`, not doc

**Comment by cannoneyed** [comment]
Hmmm not really sure why you'd have any issues with the `prepublish` command, seems like maybe the `in-publish` module isn't installed?

**Comment by cannoneyed** [comment]
Thank you, I'll try to add live-reload dev mode for the demos later tonight 🚀

**Pull request merged**